### PR TITLE
[Avro] Deprecate Avro API classes in "sdks/java/core"

### LINKED ISSUE #24749: [Task]: Deprecate Avro API classes in "sdks/java/core"
### What needs to happen?

Once all Beam modules start to use `sdks/java/extensions/avro` we need to deprecate Avro API classes in `sdks/java/core` but keep them for a while for a sake of backward compatibility. 

Affected classes:

- org.apache.beam.sdk.coders.AvroCoder
- org.apache.beam.sdk.coders.AvroGenericCoder
- org.apache.beam.sdk.io.AvroIO
- org.apache.beam.sdk.io.AvroSource
- org.apache.beam.sdk.io.DynamicAvroDestinations
- org.apache.beam.sdk.io.AvroSchemaIOProvider
- org.apache.beam.sdk.io.AvroSink
- org.apache.beam.sdk.io.SerializableAvroCodecFactory (package private)
- org.apache.beam.sdk.io.ConstantAvroDestination (package private)
- org.apache.beam.sdk.schemas.utils.AvroUtils
- org.apache.beam.sdk.schemas.utils.AvroByteBuddyUtils (package private)
- org.apache.beam.sdk.schemas.utils.ConvertHelpers
- org.apache.beam.sdk.schemas.AvroRecordSchema

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by aromanenko-dev**
[!!!] It must be merged ONLY AFTER #24992 is merged. [!!!]

All Beam modules are migrated to use `sdks/java/extensions/avro`. So it's needed to deprecate Avro API classes in `sdks/java/core` but keep them for a while for a sake of backward compatibility.

Affected classes:

- org.apache.beam.sdk.coders.AvroCoder
- org.apache.beam.sdk.coders.AvroGenericCoder
- org.apache.beam.sdk.io.AvroIO
- org.apache.beam.sdk.io.AvroSource
- org.apache.beam.sdk.io.DynamicAvroDestinations
- org.apache.beam.sdk.io.AvroSchemaIOProvider
- org.apache.beam.sdk.io.AvroSink
- org.apache.beam.sdk.io.SerializableAvroCodecFactory (package private)
- org.apache.beam.sdk.io.ConstantAvroDestination (package private)
- org.apache.beam.sdk.schemas.utils.AvroUtils
- org.apache.beam.sdk.schemas.utils.AvroByteBuddyUtils (package private)
- org.apache.beam.sdk.schemas.utils.ConvertHelpers
- org.apache.beam.sdk.schemas.AvroRecordSchema

Fixes #24749

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by aromanenko-dev** [comment]
R: @mosche 
PTAL

**Comment by aromanenko-dev** [comment]
CC: @reuvenlax @kennknowles @lukecwik

**Comment by aromanenko-dev** [comment]
CC: @Abacn

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java [hunk 1]
```diff
@@ -107,10 +107,14 @@
  * org.apache.beam.sdk.transforms.GroupByKey} operations.
  *
  * @param <T> the type of elements handled by this coder
+ * @deprecated Avro classes are deprecated to use in module <code>beam-sdks-java-core</code> and
```
Comment (mosche) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java:110]: Nitpick... maybe like below? Though not important
```suggestion
 * @deprecated Avro related classes are deprecated in module <code>beam-sdks-java-core</code> and
```

File: CHANGES.md [hunk 2]
```diff
@@ -88,6 +88,7 @@
 ## Deprecations
 
 * X behavior is deprecated and will be removed in X versions ([#X](https://github.com/apache/beam/issues/X)).
+* Avro classes are deprecated to use in module `beam-sdks-java-core` and will be eventually removed ([#24749](https://github.com/apache/beam/issues/24749)).
```
Comment (mosche) [inline CHANGES.md:91]: Nitpick... maybe like below? Though not important
```suggestion
 * @deprecated Avro related classes are deprecated in module <code>beam-sdks-java-core</code> and
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java (hunk 1 above)
Comment (mosche) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java:110]: Please also mention the new import that should be used instead to make the migration as easy as possible. Same for all below...

**Comment by aromanenko-dev** [comment]
@mosche Thanks for review, I addressed your comments, ptal

**Comment by aromanenko-dev** [comment]
retest this please

**Review by mosche — APPROVED** [review]
Thanks for adding the migration notes @aromanenko-dev 💯

**Comment by aromanenko-dev** [comment]
retest this please

**Review by Abacn — APPROVED** [review]
Thanks

**Comment by aromanenko-dev** [comment]
Failed checks are not related to this change

**Pull request merged**